include/aws/common/system_info.h

```diff
@@ -21,8 +22,48 @@ struct aws_cpu_info {
     bool suspected_hyper_thread;
 };
 
+struct aws_system_environment;
+
 AWS_EXTERN_C_BEGIN
 
+/**
+ * Allocates and initializes information about the system the current process is executing on.
+ * If successful returns an instance of aws_system_environment. If it fails, it will return NULL.
+ */
+AWS_COMMON_API
+struct aws_system_environment *aws_system_environment_load(struct aws_allocator *allocator);
+
+AWS_COMMON_API
+void aws_system_environment_destroy(struct aws_system_environment *env);
+
+/**
+ * Returns the virtualization vendor for the specified compute environment, e.g. "Xen, Amazon EC2, etc..."
+ *
+ * The return value may be empty and in that case no vendor was detected.
+ */
+AWS_COMMON_API
+struct aws_byte_cursor aws_system_environment_get_virtualization_vendor(const struct aws_system_environment *env);
```

Comment:
With this API, there's no way to differentiate "unknown because none of my heuristics matched" from "unexpected error while trying to determine (e.g. IMDS flakiness)"

If you returned `struct aws_string *` you could return empty string for unknown, and NULL-with-error if "unexpected error"

but then, the error would have happened earlier, during `aws_system_environment_load()`, so you'd need to store the reason for failure, and raise it when the appropriate getter is called

Reply:
there's no IMDS flaky ness. There's no AWS here, it's just generic, read crap off of the OS as best you can.

There's no handleable error. You either have the info or you don't.